### ISSUE TITLE
Permission when creating shows/season folders

### ISSUE THREAD

**Issue by udochrist** [SiCKRAGETV/sickrage-issues#2969]
I'm running into a problem with how the permissions are assigned to freshly created folders for shows/seasons and PLEX.

i understand that Sickrage creates folders and then updates the permission to that of the parent folder. This is working o.k.
i cannot find any reference about the permissions of the initial creation of a folder though. Could it be that these folders are created to be accessed by the user-only?

SickRage and Plex belong to the same user-group on my server. However each time SickRage creates a new folder the Plex gets triggered by the filesystem update but fails due to permissions on that folder.

Sep 22, 2015 03:20:51 [0xe9d59b70] ERROR - [Notify] Failed to add watch for /volume1/Media/TV Shows/BlaBlaBla/Season 01 (13: Permission denied)

 When i later click to manually update my library in Plex there are no more permission problems.

would it therefore be possible (by adding a umask) to get control of the permission bits used when initially creating folders?

**Comment by neoatomic** [comment]
https://github.com/SiCKRAGETV/sickrage-issues#submitting-a-bugissue-ticket

**Comment by udochrist** [comment]
Its actually more of a Feature request than a Support question. In SickBeard.py the default umask is changed to 077. i would like to be able to Control that value from the config screen

**Comment by neoatomic** [comment]
Dont think thats a good idea. @miigotu ?

**Comment by miigotu** [comment]
```python
>>> os.umask(0) and int('077', 8)
0
```

It is setting a umask of 0. Your problem lies somewhere else.

**Comment by udochrist** [comment]
No. That's the place if you use 007 instead of the 077 then the group bits are kept as the are. The numbers are subtracted from the permission fwiw. In the end this mask is not really relevant since after creation the folders are being set to the parents folder permissions anyway. I've tested 007 and this fixes my plex issue

**Comment by miigotu** [comment]
Its anding with 077, and calling umask twice. Its kinda hard to follow what is going on there.

Are you saying the mask should be 0o022 or 0o027?

**Comment by udochrist** [comment]
Thanks for looking into this.. The problem has caused me a lot of headache already and I always keep a modified Python file as a fix which gets overwritten with each update. 

Longer description of the issue/remedy:
Umask takes away the permissions when you create a file. So if you start with 777 as the permission and apply the umask 000 you will keep the 777 if you use a umask of 022 you would end up with 755 (rwxr-xr-x).

What the code does in sickbeard.py is to retrieve the current umask by setting it to zero, then the last parts (group and world) are completely removed by the 077 keeping only the first octet so only the current user can read the files and then the umask is set back keeping the owner part. (I find it a bit weird to do it that complex way instead of simply using 700, but whatever)

If in this case you start with a rather common UNIX umask of 022 (meaning group and world readable) you end with 077 meaning only the user can read files. On my server plex and sickrage run under a different uid which means that in this case plex cannot access the files and directories created by sickrage. The both users (for me) share the group "media" hence plex is immediately able to see sickrages files if I keep the group permissions (I.e. 007 instead of 077) in the code. Funny thing is that after sickrage creates a folder only readable by itself it immediately changes the permissions to that of the parent folder. Plex however gets triggered directly after the creation with the still user only limited folder access and fails to scan & watch the folder. 
When it comes to permissions it's ok to be cautious however this way automatic library update in plex cannot work. Therefore my idea would be to either leave out the umask stuff completely since the admin could set the desired umask before starting sickrage or to provide the default umask in the user editable config and use that  umask directly instead of subtracting 077 from the current umask as its done right now which is a tad confusing.

**Comment by miigotu** [comment]
I would like to decide on a good umask to use, without making it configurable, but a config.ini only option is not out of the question. I agree this is an issue and should be addressed.

Imo 022 final umask is not a good option because plex (or sabnzbd etc) who are just in the same group would not be able to write metadata or other media to the shows.
This also rules out 027.

umask 0 might too permissive, but it might be the best bet also

007 is a potential option as it would allow group to execute and block all others (but should others be denied rw?)

I think final umask of 011 or even 111 might be a good option. Thoughts?
What potential flaws could you see from leaving it out altogether?

**Comment by udochrist** [comment]
Then I'd go for "and-ing" with 007 instead of the current 077. This immediately fixed my plex vs sickrage auto update issues. I prefer the config ini solution however since with any defaults you cannot make it right in complex configs where the user wants a special setup. 

My opinion is: either don't touch umasks at all in a program or if you want to give the user the full control let him specify the umask directly.

**Comment by udochrist** [comment]
Not being able to specify the umask and have a good value as a default will render users w/o any UNIX skills and start/stop script knowledge helpless.  That was the rationale of my proposal to have the config option in the gui.

**Comment by miigotu** [comment]
"will render users w/o any UNIX skills and start/stop script knowledge helpless"
exactly the reason I DONT want it configurable. I want it to just be right. If that means setting a better one like 011 cool, but if users who have no knowledge can change this then I will have an assload of issues that are difficult to diagnose because of people changing it to something that makes their setup not work.

Are you prepared to take on that extra work of explaining umasks and ANDing to 8k+ (many amateur) users? lol

007 is too restrictive, since media centers on different systems arent in the same group and may need to write to a show folder.

I think 011 would not be ok because updates are ran through a popen git after umask is set and created files would not be x for group (autoProcessTV/nzbtomedia/etc)

So that leaves 001 and 0. I feel there should be at least SOME masking to prevent non-user/non-group from running any of these files and borking an install or something by creating say a DB them with under the wrong user.

If 001 sounds sufficient I will make that change.

**Comment by udochrist** [comment]
I understand your point.. However: If there are current issues with this I did not find them in the tracker before filing this request so I can't really comment how many people will try to change the values if every thing works. If there are a lot of issues with this I would not mess with it at all in the code but simply leave the umask as it is when sr gets started. 

The way couchpotato does this is the easiest to understand I think. Drop umasks but specify the permissions to use when creating files/folders. Should be easy since you already do this after creating files/folders. Just the creation needs to be changed then. Attached is a screenshot of couchpotatos option which is hidden under "advanced"

**Comment by miigotu** [comment]
Either way, the logic in the code there in SR is wrong.

When umask starts as 0 it is never masked with that logic:
```python
>>> 0o000 & 0o077
0
>>> 0o000 and 0o077
0
>>> 0o000 | 0o077
63
>>> 0o077
63
```

I'm going to think more about this before I change it.

**Comment by miigotu** [comment]
Did some research and decided to go with the normal convention for daemonizing and set umask 0. Since we chmod as parent anyways this wont have any negative effects.
